### PR TITLE
upstream (ticdc): complete upstream and add unit tests

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: ref #5288

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

### Release note 

```release-note

None
```

### CODE REVIEW

File: pkg/upstream/upstream.go [hunk 1]
```diff
@@ -58,13 +61,20 @@ type Upstream struct {
 	PDClock     pdutil.Clock
 	GCManager   gc.Manager
 
+	// holder count of this upstream
+	hc int32
+	// record the time when Upstream.hc becomes zero.
+	idleTime time.Time
+	// mu should be lock when Upstream.hc or Upstream.idleTime need to be changed.
+	mu sync.Mutex
```
Comment (CharlesCheung96) [inline pkg/upstream/upstream.go:69]: Perhaps these three fields should be combined into single struct.
ref:https://github.com/pingcap/tiflow/pull/5264/files#diff-cfe8ccb173f8b93acc1b97cc5825fdb565d539634714fadffacdf18ea4bc8b26L145-R33

File: pkg/upstream/manager.go [hunk 2]
```diff
@@ -47,41 +49,44 @@ func NewManager(ctx context.Context) *Manager {
...
 	default:
 	}
-
-	if _, ok := m.ups.Load(clusterID); ok {
+	if _, ok := m.ups.Load(upstreamID); ok {
 		return nil
 	}
 	securityConfig := config.GetGlobalServerConfig().Security
-	up := newUpstream(clusterID, pdEndpoints, securityConfig)
+	up := newUpstream(upstreamID, pdEndpoints, securityConfig)
```
Comment (sdojjy) [inline pkg/upstream/manager.go:70]: make securityConfig as the function parameter

**Review by sdojjy — APPROVED** [review]

File: pkg/txnutil/gc/testing.go [hunk 3]
```diff
@@ -36,3 +36,5 @@ func (m *MockPDClient) UpdateServiceGCSafePoint(ctx context.Context, serviceID s
 func (m *MockPDClient) GetTS(ctx context.Context) (int64, int64, error) {
 	return oracle.GetPhysical(time.Now()), 0, nil
 }
+
+func (m *MockPDClient) Close() {}
```
Comment (CharlesCheung96) [inline pkg/txnutil/gc/testing.go:40]: Why do we need an empty close method here?

File: pkg/upstream/upstream.go [hunk 4]
```diff
@@ -198,3 +220,42 @@ func (up *Upstream) IsNormal() bool {
...
+
+func (up *Upstream) unhold() {
+	up.hcMu.mu.Lock()
+	defer up.hcMu.mu.Unlock()
+	up.hcMu.hc--
+	if up.hcMu.hc < 0 {
+		log.Panic("upstream's hc should never less than 0", zap.Uint64("upstreamID", up.ID))
+	}
+	if up.hcMu.hc == 0 && up.hcMu.idleTime.IsZero() {
+		up.hcMu.idleTime = time.Now()
```
Comment (CharlesCheung96) [inline pkg/upstream/upstream.go:232]: IdleTime should always be updated when `up.hcMu.hc == 0`.

File: pkg/upstream/upstream.go [hunk 5]
```diff
@@ -198,3 +220,42 @@ func (up *Upstream) IsNormal() bool {
...
+
+func (up *Upstream) Release() {
+	up.unhold()
+}
+
+// return true if this upstream idleTime reachs maxIdleDuration.
+func (up *Upstream) shouldClose() bool {
+	up.hcMu.mu.Lock()
+	defer up.hcMu.mu.Unlock()
+	if !up.hcMu.idleTime.IsZero() && time.Since(up.hcMu.idleTime) >= maxIdleDuration {
```
Comment (CharlesCheung96) [inline pkg/upstream/upstream.go:257]: Should check if `hc==0` here, rather than idleTime.

File: pkg/upstream/upstream.go [hunk 6]
```diff
@@ -198,3 +220,42 @@ func (up *Upstream) IsNormal() bool {
...
+	up.hcMu.hc--
+	if up.hcMu.hc < 0 {
+		log.Panic("upstream's hc should never less than 0", zap.Uint64("upstreamID", up.ID))
+	}
+	if up.hcMu.hc == 0 && up.hcMu.idleTime.IsZero() {
+		up.hcMu.idleTime = time.Now()
+	}
+}
+
+func (up *Upstream) hold() {
```
Comment (CharlesCheung96) [inline pkg/upstream/upstream.go:236]: The names of `hold` and `hc` are somewhat confusing, perhaps we can find a more appropriate name?

File: pkg/txnutil/gc/testing.go (hunk 3 above)
Comment (asddongmen) [reply pkg/txnutil/gc/testing.go:40]: To implement PDClient interface.

File: pkg/upstream/upstream.go (hunk 5 above)
Comment (asddongmen) [reply pkg/upstream/upstream.go:257]: No, `hc` can be `0` when a upstream is created but is not hold by a `changefeed` or `processor`.

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**